File: go/fixture/go-mini/internal/models/tenant.go
package models

import (
	"errors"
	"fmt"
)

const maxTenantLen = 32

// Tenant identifies the owner of a device fleet. A Tenant can only be
// obtained through ParseTenant, so every value in circulation is already
// non-empty, lowercase, and at most 32 characters long; nothing downstream
// needs to check it again.
type Tenant struct {
	id string
}

// ParseTenant validates a raw tenant identifier once, at the boundary.
func ParseTenant(raw string) (Tenant, error) {
	if raw == "" {
		return Tenant{}, errors.New("tenant: empty")
	}
	if len(raw) > maxTenantLen {
		return Tenant{}, fmt.Errorf("tenant %q: longer than %d characters", raw, maxTenantLen)
	}
	for _, r := range raw {
		if !isTenantRune(r) {
			return Tenant{}, fmt.Errorf("tenant %q: want lowercase letters, digits, or '-'", raw)
		}
	}
	return Tenant{id: raw}, nil
}

// ID returns the validated identifier for use in storage keys and logs.
func (t Tenant) ID() string {
	return t.id
}

func isTenantRune(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '-'
}

File: py/fixture/py-mini/internal/models/tenant.py
"""Tenants: the owners of device fleets."""

from dataclasses import dataclass
from typing import Self

_MAX_TENANT_LEN = 32
_TENANT_CHARS = frozenset("abcdefghijklmnopqrstuvwxyz0123456789-")


@dataclass(frozen=True)
class Tenant:
    """Identifies the owner of a device fleet.

    A Tenant can only be obtained through Tenant.parse, so every value in
    circulation is already non-empty, lowercase, and at most 32 characters
    long; nothing downstream needs to check it again.
    """

    _id: str

    @classmethod
    def parse(cls, raw: str) -> Self:
        """Validate a raw tenant identifier once, at the boundary."""
        if not raw:
            raise ValueError("tenant: empty")
        if len(raw) > _MAX_TENANT_LEN:
            raise ValueError(f'tenant "{raw}": longer than {_MAX_TENANT_LEN} characters')
        for r in raw:
            if not _is_tenant_char(r):
                raise ValueError(f"tenant \"{raw}\": want lowercase letters, digits, or '-'")
        return cls(raw)

    @property
    def id(self) -> str:
        """Return the validated identifier for use in storage keys and logs."""
        return self._id


def _is_tenant_char(r: str) -> bool:
    return r in _TENANT_CHARS
